src/FormLayout/Generic.php

```diff
@@ -35,43 +36,67 @@ class Generic extends View
     public $inline = null;
 
     /**
-     * Places field inside a layout somewhere.
+     * Places field inside a layout somewhere. Should be called
+     * through $form->addField().
      *
-     * @param \atk4\ui\FormField\Generic|array $field
-     * @param array|string                     $args
+     * @param string|null              $name
+     * @param array|string|object|null $decorator
+     * @param array|string|object|null $field
      *
      * @return \atk4\ui\FormField\Generic
      */
-    public function addField($field, $args = [])
+    public function addField(string $name, $decorator = null, $field = null)
     {
-        if (is_string($args)) {
-            $args = ['caption' => $args];
-        } elseif (is_array($args) && isset($args[0])) {
-            $args['caption'] = $args[0];
-            unset($args[0]);
+        if (!is_string($name)) {
+            throw new Exception(['Format for addField now require first argument to be name']);
         }
 
-        /*
-        if (isset($args[1]) && is_string($args[1])) {
-            $args[1] = ['ui'=>['caption'=>$args[1]]];
+        if (!$this->form->model) {
+            $this->form->model = new \atk4\ui\misc\ProxyModel();
         }
-         */
 
-        if (is_array($field)) {
-            $field = $this->form->fieldFactory(...$field);
-        } elseif (!$field instanceof \atk4\ui\FormField\Generic) {
-            $field = $this->form->fieldFactory($field);
+        if (is_string($field)) {
+            $field = ['type'=>$field];
         }
 
-        if (isset($args['caption'])) {
-            $field->field->caption = $args['caption'];
+        if ($name) {
+            $existingField = $this->form->model->hasElement($name);
         }
 
-        if (isset($args['width'])) {
-            $field->field->ui['width'] = $args['width'];
+        if (!$existingField) {
+            // Add missing field
+            if ($field) {
+                $field = $this->form->model->addField($name, $field);
+            } else {
+                $field = $this->form->model->addField($name);
+            }
+        } elseif (is_array($field)) {
+            // Add properties to existing field
+            $existingField->setDefaults($field);
+            $field = $existingField;
+        } elseif (is_object($field)) {
+            throw new Exception(['Duplicate field', 'name'=>$name]);
+        } else {
+            $field = $existingField;
         }
 
-        return $this->_add($field, ['name'=>$field->short_name]);
+        if (is_string($decorator)) {
+            $decorator = $this->form->decoratorFactory($field, ['caption'=>$decorator]);
```

Comment:
I am myself unsure about this code. IIRC, Decorator should indicate type and third argument may contain caption.